Base: Correlate Mesh & Particle Groups

ax3l opened the pull request:
## Description

This allows to add an optional attribute to particle groups and mesh records to cross-correlate projections and probes.

*Implements issue:* #238

## Affected Components

- [x] `base`

## Logic Changes

TODO: (write this)

## Writer Changes

None

## Reader Changes

TODO: (write this)

*What would a reader need to change? Link implementation examples!*

- [x] `openPMD-validator`: 
- [ ] `openPMD-viewer`: 
- [ ] `VisualPIC`: https://github.com/AngelFP/VisualPIC
- [x] `yt`: none
- [x] `VisIt`: none
- [x] `converter`: none
- [x] `openPMD-api`: none

## Data Updater

None

ax3l committed:
Base: Correlate Mesh & Particle Groups

This allows to add an optional attribute to particle groups and
mesh records to cross-correlate projections and probes.

AngelFP reviewed (commented):
Thanks for this nice addition, @ax3l . This would definitely simply some things in VisualPIC!

I have one question though: the implementation here works well to link mesh records to particle groups, but this means that both have to exist. However, it can happen sometimes that, for example, the simulation output contains the charge density field associated to a particle species, but does not contain the species itself as particle group. In this case, the mesh is therefore not linked to a particle group, but simply to a species in the simulation. Would this be a problem here?

ax3l commented:
That's a good point. I think we can add a note for implementers to clarify.

ax3l commented on STANDARD.md:
```diff
@@ -457,6 +457,20 @@ The following attributes must be stored with each `scalar record` and each
+    - note:
+        this is often used in graphical user-interfaces to group projected meshes of particle distributions in a user-friendly manner;
+        examples are particle densities and phase space projections
+    - examples:
```
```suggestion
    - note:
        when reading this back, the particle groups listed here might not be present in some steps or are even skipped in the whole openPMD series by the writer
    - examples:
```

ax3l commented on STANDARD.md:
```diff
@@ -597,6 +611,20 @@ patch order:
+    - note:
+        this is often used in graphical user-interfaces to group meshes that are sampled with particles in a user-friendly manner;
+        examples are particle probes as well as non-uniform and curved samples from large meshes
+    - examples:
```
```suggestion
    - note:
        when reading this back, the mesh records listed here might not be present in some steps or are even skipped in the whole openPMD series by the writer
    - examples:
```

ax3l committed:
Clarify: Naming Skipped Allowed

Clarify that referencing particle groups and mesh records that are
skipped in output is ok.

AngelFP commented:
Looks good, thanks a lot!

> It's still really useful to reference a particle group name for meshes like `<species>_chargeDensity`, even if the particle output is disabled for an openPMD series.

That's also a nice convention. I'll keep it in mind for when I have to implement it in Wake-T.
